contributors/design-proposals/security-model.md

```diff
@@ -0,0 +1,190 @@
+of "foo". The service's backends would also be limited to pods owned by a group.
+
+1. All resources would be owned.
+2. Scoping down permissions is simple with control over both object and relationship access\*\*.
+3. These owners could derive from an identity management system, so the onus of
+enforcement isn't on the cluster admin.
+
+A true ownership model solves the disadvantages of namespace-based ownership.
+But it's not clear that the improvements are worth the effort of implementation.
+
+\* Group is only used as an example. The system could be built to support
+users, roles, service accounts, or any other identifiable entity.
+
+\*\* It's also reasonable to control relationships with a reflective or membership
+constraint. This simplifies permissions management at the expense of
+flexibility, and gives no real improvement over problem (2) with namespaces.
+
+# Workloads
+
+The other face of Kubernetes exposed to customers is the workload environment.
+This section focuses on protecting workloads from the environment, and the
+environment from workloads.
+
+## Scheduled isolation
+
+Many users have use cases for scheduled isolation. This is where Kubernetes
+separates sensitive workloads from each other.
+
+1. A sensitive application might not tolerate other workloads on the same node.
+2. A user may bring a dedicated node.
+3. An application may need to run privileged containers.
+
+Both affinity and taints can achieve this today, but neither is protected. Both
+affinities and tolerations are specific on the "PodSpec", so anyone with "create
+pod" permissions can circumvent the mechanism.
+
+### Dedicated nodes
+
+Dedicated nodes require scheduling protection, because A's safety shouldn't be
+dependent on B's goodwill. Taint ACLs would suffer from the same problem as
+label ACLs. However, the "ownership" model can be applied successfully.
+
+For instance, a special "allowed-namespaces" taint would forbid pods outside of
+authorized namespaces. The pods would still be responsible for selecting the
+nodes.
+
+Alternatively, membership in a group, or a role, could be used to permit or
+forbid execution on a node. This might make more sense than forcing namespaces
+onto nodes, but probably requires the intentional ownership model.
+
+### Privileged pods
+
+A PodSecurityPolicy can prevent users from creating privileged containers.
+However, some use cases may be that privileged pods can only be created on
+dedicated nodes.
+
+For instance, a game server may want to run with host networking. These will
+need to be created with a privileged PSP. But many workloads may not tolerate
+sharing a host with privileged user pods.
+
+1. Privileged containers could taint a node.
+2. A cluster-level policy could restrict privileged containers to dedicated
+   nodes.
+3. A "NodeSecurityPolicy" could be introduced, which could forbid privileged
+   user containers.
```

Comment:
I meant to exclude controllers with "user containers". If that's unclear, can you suggest another phrasing?

Are we considering nodes that don't trust the apiserver? It sounds fine for nodes to be the final authority on what they execute, and therefore reject pods.

1. I don't see why the cluster's control plane shouldn't know these restrictions for placement. If 90% of my fleet wants to reject privileged containers, then I would have quite a few mishaps before I found a stable home.

1. Nodes could still be the authority. For instance, the kubelet could pull a "NodeSecurityPolicy" out of a local yaml file, and then write it back to the apiserver. The apiserver will respect the node's policy, but if they get out of sync, then the node will still reject forbidden workloads.